docs/proposals/loadbalancing.md

```diff
@@ -0,0 +1,338 @@
+
+__Ingress paths__
+
+1. Routes: A pathmap is conceptually inline with how people view the problem, and allows us to perform atomic switch over for blue green deployments. We can still expose a "route like" update interface so users don't have send the entire pathmap to change a single url endpoint.
+
+2. GCEs UrlMap: This resource is way too confusing. It contains a list of host rules each of which can point to a list of path matchers, each of which has a list of path regexes. Users end up bearing the brunt of this complexity.
+
+3. Load balancer resource: This resource would encapsulate all the information in claims and ingress paths. It will get unmanageable in deployments where a single global loadbalancer is claiming all ingress paths. This also doesn't provide a clean abstraction between acquirigin an IP and loadbalancing (the user impedance mismatch problem), while in theory we can have a non-loadbalancer fulfill a claim for an ip, or prepopulate a claim with a known ip just to have services reference it.
+
+__Claims__
+
+1. Have a global loadbalancer claim Ingress paths till it can claim no more: If we decide that users don't get to request new public ips, we could just dissolve the claim into the ingress paths and make the loadbalancer controllers smarter.
+  * Users want to use an ip already in DNS for the service
+  * The controllers need to become smarter of cloudprovider limitations like only using a single cert per loadbalancer ip
+  * Detecting when a single loadbalancer is saturated can be hard
+
+### Issues
+
+What follows is a digest of some of the problems brought up in various github issues. Expect these to make it into the proposal above in due time.
+
+#### How does this work at L4?
+
+People choose L4 for a few different reasons:
+- Cost
+- Latency
+- Throughput of loadbalancer is the bottleneck
+- Cloud provider limitations
+- Want SSL passthrough or SNI switch for multi-certs
+- Need client IP and cannot set loadbalancer proxy as default gateway
+
+Point being, we can't reliably infer all these use case with Service.Type=LoadBalancer or the absence of a UrlMap in the IngressPoint (because they might want L7 for session persistence). Instead, anything serving ingress traffic needs an ingress path (which is why it isn't called urlMap). An L4 ingress path might look like:
+
+```yaml
+apiVersion: v1
+kind: IngressPoint
+metadata:
+  name: l4ingress
+  type: network
+spec:
+  host: www.example.com
+  tlsMode: Termination
+  secret: foosecret
+status:
+  loadBalancer:
+    ingress:
+    - ip: 104.1..
+```
+
+type: network with a pathmap is a validation error.
+This adds a slight burden on users, since what was previously a one line change to expose a service now requires a new resource.
```

Comment:
Or you at least need a port map for L7 from a single or set of hosts.  In
either way it's a map.

On Tue, Sep 8, 2015 at 2:48 PM, Prashanth B <notifications@github.com>
wrote:

> In docs/proposals/loadbalancing.md
> <https://github.com/kubernetes/kubernetes/pull/12827#discussion_r38964334>
> :
>
> > +kind: IngressPoint
> > +metadata:
> > +  name: l4ingress
> > +  type: network
> > +spec:
> > +  host: www.example.com
> > +  tlsMode: Termination
> > +  secret: foosecret
> > +status:
> > +  loadBalancer:
> > +    ingress:
> > +    - ip: 104.1..
> > +```
> > +
> > +type: network with a pathmap is a validation error.
> > +This adds a slight burden on users, since what was previously a one line change to expose a service now requires a new resource.
>
> For now, I'm thinking of l7 and l4 independently. L4 will be through
> service type, l7 through ingress point (because you need a pathmap for l7
> anyway).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12827/files#r38964334>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift